### PR TITLE
OSSM-1968 Move 2.3 CNI container into separate DaemonSets

### PR DESCRIPTION
**Pull request by yxun**
Signed-off-by: Yuanlin <yuanlin.xu@redhat.com>

This PR replaces the cni template daemonset.yaml with separated ones: daemonset-v2.0.yaml, daemonset-v2.1.yaml, daemonset-v2.2.yaml and daemonset-v2.3.yaml

I use the existing cni.supportedReleases values to filter a SMCP controller version, because there is no direct mapping between cni bootstrap and a controller spec version. 

In upgrade case, the new boostrap code will check and delete the existing "istio-node" and "istio-cni-node" DaemonSets if they exist.

### CODE REVIEW

**Comment by yxun** [comment]
@jewertow @luksa 
PR ready for a review but not ready for merge. I will test my operator image on an OCP later today. Thanks,

**Comment by yxun** [comment]
I will convert it from Draft to ready for review after fixing CI issues. The CI issues are waiting for CI team approval in another repo PR.

**Review by dgn — COMMENTED** [review]
Wasn't the intention to only install those DaemonSets that are required? In that case, I would say we extend `InstallCNI()` to also pass the SMCP version (it's called from within the reconciler, so that should be easy), which can then be passed on to `internalRenderCNI()` so we don't always render all DaemonSets

**Comment by jewertow** [comment]
I think that the goal of this issue was to install a separated daemonset only for 2.3. Previous versions should work as before.

**Comment by luksa** [comment]
> I think that the goal of this issue was to install a separated daemonset only for 2.3. Previous versions should work as before.

+1

Although it would be nice to have separate daemonsets for each version, this would be a bigger change and could introduce upgrade issues. That was one of the reasons we decided to keep one daemonset for all the older versions and only introduce a new one for 2.3 (when we do 2.4, we'll add another daemonset, and so on...)

**Comment by yxun** [comment]
> I think that the goal of this issue was to install a separated daemonset only for 2.3. Previous versions should work as before.

Ok, that makes this PR easier. i will update it today. 

> Wasn't the intention to only install those DaemonSets that are required? In that case, I would say we extend InstallCNI() to also pass the SMCP version (it's called from within the reconciler, so that should be easy), which can then be passed on to internalRenderCNI() so we don't always render all DaemonSets

Yes, totally agrees. I spent some time trying to pass the SMCP controller version but it turned out to be a bigger change than I expected. 
After 2.3 release, I can read more and think about a way of extending that. Thanks.

**Comment by yxun** [comment]
unit test passed in my local laptop and the new job but failed in our prow job. is there anything changed in our prow ?

File: pkg/bootstrap/cni_test.go
Comment (jewertow) [inline pkg/bootstrap/cni_test.go:34]: Did this order matter?

Comment (jewertow) [inline pkg/bootstrap/cni_test.go]: Does order of these names matter?

Comment (yxun) [reply pkg/bootstrap/cni_test.go]: yes, it matters now. because it finds the v2.3 one and append it first

File: pkg/bootstrap/cni_test.go
Comment (jewertow) [inline pkg/bootstrap/cni_test.go]: To be honest, I don't understand what's going on here. Shouldn't we just remove case "v2.3 only"?

File: manifests-maistra/2.3.0/maistraoperator.v2.3.0.clusterserviceversion.yaml
Comment (jewertow) [inline manifests-maistra/2.3.0/maistraoperator.v2.3.0.clusterserviceversion.yaml]: We usually avoid updating `createdAt`, but I am not sure if it is a problem before we release a particular version...?

File: pkg/bootstrap/cni_test.go
Comment (jewertow) [reply pkg/bootstrap/cni_test.go]: Thanks!

File: pkg/bootstrap/cni_test.go
Comment (yxun) [reply pkg/bootstrap/cni_test.go]: Because in v2.3 case, we will have two daemonsets. One from our original daemonset.yaml and one from the new template daemonset-v2.3.yaml.

When we create a v2.3 SMCP, supportedVersion only has [v2.3] value, so the daemonset.yaml will not include any container. That's why 
```
unstructured.NestedSlice(resource.UnstructuredContent(), "spec", "template", "spec", "containers") 
```
returns nil and found is false in that case.

However, if I change the rest of the if statements in the daemonset.yaml, that will break existing clusters in upgrade.
In that case, only removing the v2.3 container section in the daemonset.yaml template will be safe in upgrade.

File: manifests-maistra/2.3.0/maistraoperator.v2.3.0.clusterserviceversion.yaml
Comment (yxun) [reply manifests-maistra/2.3.0/maistraoperator.v2.3.0.clusterserviceversion.yaml]: That's in the second commit. There are new changes generated from the `make gen` so I think I need to update the manifest date. If that's not good, I can checkout the original values

File: pkg/bootstrap/cni_test.go
Comment (jewertow) [reply pkg/bootstrap/cni_test.go]: Ok, thanks! Now I understand. Could you add more precise comment?

Comment (yxun) [reply pkg/bootstrap/cni_test.go]: sure, i will add more

File: manifests-maistra/2.3.0/maistraoperator.v2.3.0.clusterserviceversion.yaml
Comment (jewertow) [reply manifests-maistra/2.3.0/maistraoperator.v2.3.0.clusterserviceversion.yaml]: Yes, but manifest did not change itself. @jwendell could you confirm?

**Review by jewertow — COMMENTED** [review]
Generally looks good to me. But before I approve, I would like to test it locally, ok?

File: manifests-maistra/2.3.0/maistraoperator.v2.3.0.clusterserviceversion.yaml
Comment (yxun) [reply manifests-maistra/2.3.0/maistraoperator.v2.3.0.clusterserviceversion.yaml]: It changed. I talked to jonh in a slack thread
https://coreos.slack.com/archives/C01R9E3SFEV/p1664464705920469

**Comment by luksa** [comment]
@yxun when I create a 2.3 SMCP, the operator installs both the `istio-cni-node` and `istio-cni-node-v2.3`. The goal was for the operator to only install the 2.3 daemonset. The operator should install the old daemonset only when an SMCP v2.0, v2.1 or v2.2 is created.

**Comment by yxun** [comment]
> @yxun when I create a 2.3 SMCP, the operator installs both the `istio-cni-node` and `istio-cni-node-v2.3`. The goal was for the operator to only install the 2.3 daemonset. The operator should install the old daemonset only when an SMCP v2.0, v2.1 or v2.2 is created.

@luksa Hi Marko, thanks. I updated this PR. And the two tests cover your comment above:
- "v2.3 only" verifies the correct cni container is rendered
https://github.com/maistra/istio-operator/blob/f258ef451cfef2e66aaf30c35bd8c5bea8aebdcc/pkg/bootstrap/cni_test.go#L57

The reason why I moved the "names := []string{}" outside of the for loop is for matching the ""Default Supported Versions" test case.

- ""bootstrap-clean-install-cni-no-errors" verifies the correct daemonset is able to be installed and the other one is not 
https://github.com/maistra/istio-operator/blob/f258ef451cfef2e66aaf30c35bd8c5bea8aebdcc/pkg/controller/servicemesh/controlplane/integration_test.go#L205
https://github.com/maistra/istio-operator/blob/f258ef451cfef2e66aaf30c35bd8c5bea8aebdcc/pkg/controller/servicemesh/controlplane/integration_test.go#L213

**Comment by luksa** [comment]
@yxun, I'm not talking about `supportedVersions`. One of the acceptance criteria in the Jira is the following:

> - operator deploys the correct DaemonSet version when an SMCP of a particular version (2.3 and forward) is deployed

So, you'll need to pass the SMCP version to `InstallCNI()` in [1] and make sure that either the old daemonset, or the 2.3 daemonset gets created, depending on what the SMCP version is.

[1] https://github.com/Maistra/istio-operator/blob/36327f8a0e3f832d1c3a02145de16e4aa4cd268c/pkg/controller/servicemesh/controlplane/reconciler.go#L212

**Comment by yxun** [comment]
@luksa I misunderstood @dgn and your comments before. I have updated the change.

File: resources/helm/overlays/istio_cni/templates/daemonset-v2.3.yaml
Comment (luksa) [inline resources/helm/overlays/istio_cni/templates/daemonset-v2.3.yaml]: Why check if `supportedReleases` is not nil? Shouldn't this check whether `supportedReleases` contains `v2.3`?

File: resources/helm/overlays/istio_cni/templates/daemonset.yaml
Comment (luksa) [inline resources/helm/overlays/istio_cni/templates/daemonset.yaml]: Why this check?

File: resources/helm/overlays/istio_cni/templates/daemonset-v2.3.yaml
Comment (luksa) [inline resources/helm/overlays/istio_cni/templates/daemonset-v2.3.yaml]: I'd prefer to see the `istio-cni-configmap` also split into two ConfigMaps, so that every version above 2.3 has its own ConfigMap. This should make it easier to rebase to new versions.

File: pkg/bootstrap/cni.go
Comment (luksa) [inline pkg/bootstrap/cni.go]: Is this necessary? If it is, why not just do this in one place (for example, `internalRenderCNI()`)? But I'd prefer to completely remove this and instead have the code fail. This makes it easier to reason about the code (you either pass the version when you invoke the function, or you don't... if the function uses a default value when the caller passes in nil, it's possible this will eventually lead to some weird bug).

File: resources/helm/overlays/istio_cni/templates/daemonset-v2.3.yaml
Comment (luksa) [inline resources/helm/overlays/istio_cni/templates/daemonset-v2.3.yaml]: Do we want to call this `istio-cni-node-v2-3` instead so that the name and the `k8s-app` label value are consistent?

**Review by luksa — CHANGES_REQUESTED** [review]

File: pkg/bootstrap/cni.go
Comment (yxun) [reply pkg/bootstrap/cni.go]: I found a nil case from the cni_test.go
https://github.com/maistra/istio-operator/blob/maistra-2.3/pkg/bootstrap/cni_test.go#L35

In our unit test we called the internalRenderCNI() directly. I can remove my DefaultVersion setting above and then modify the cni_test unit test code.
https://github.com/maistra/istio-operator/blob/maistra-2.3/pkg/bootstrap/cni_test.go#L72

File: resources/helm/overlays/istio_cni/templates/daemonset-v2.3.yaml
Comment (yxun) [reply resources/helm/overlays/istio_cni/templates/daemonset-v2.3.yaml]: I checked the versions code and I don't see a case that supportedReleases will be nil as long as ver.Info.MinimumSupportedVersion is not InvalidVersion
https://github.com/maistra/istio-operator/blob/maistra-2.3/pkg/controller/versions/versions.go#L85

and yes, I updated that line with 
{{- if and (has "v2.3" .Values.cni.supportedReleases) (eq .Values.cni.instanceVersion "v2.3") }}
If a future release operator will deprecate v2.3, the first if check will skip rendering that daemonset.

File: resources/helm/overlays/istio_cni/templates/configmap-v2.3.yaml
Comment (luksa) [inline resources/helm/overlays/istio_cni/templates/configmap-v2.3.yaml]: We should probably change this to just `cni_network_config`, since it's the only entry in this ConfigMap, but we can do this in a follow-up PR.

**Review by luksa — APPROVED** [review]
LGTM! Nice job.

File: resources/helm/overlays/istio_cni/templates/configmap-v2.3.yaml
Comment (yxun) [reply resources/helm/overlays/istio_cni/templates/configmap-v2.3.yaml]: I rebased the CM data entry name change in this PR. The only place we set that network config name is in:
https://github.com/maistra/istio-operator/pull/1017/files#diff-fb85c6f4f7143e3ebb7cf98f29325df00f2945710262a52c73b64d790c12ed53R64

**Comment by jewertow** [comment]
I tried to install SMCP from your branch and CNI daemon set fails with the following error:
```
2022-10-04T14:41:56.694363Z	error	Installer exits with open /host/etc/cni/multus/net.d/v2-3-istio-cni.kubeconfig.tmp.3542813523: no such file or directory
2022-10-04T14:41:56.694369Z	info	install	Cleaning up.
2022-10-04T14:41:56.694378Z	info	install	Removing binary: /host/opt/cni/bin/v2-3-istio-cni
Error: open /host/etc/cni/multus/net.d/v2-3-istio-cni.kubeconfig.tmp.3542813523: no such file or directory
```

**Comment by yxun** [comment]
> 
v2-3-istio-cni  that file should be created from the configmap. Thanks, Jaecek. I am checking if that missing file was caused by the last comment above when I renamed the 2.3 configmap data entry.

**Comment by luksa** [comment]
@yxun Why is this on hold? Let's merge it, since it needs to go into 2.3.

**Comment by luksa** [comment]
> > 
> 
> v2-3-istio-cni that file should be created from the configmap. Thanks, Jaecek. I am checking if that missing file was caused by the last comment above when I renamed the 2.3 configmap data entry.

Definitely. I tested everything before you renamed the entry and it worked fine.

**Comment by yxun** [comment]
> @yxun Why is this on hold? Let's merge it, since it needs to go into 2.3.

@luksa Jaecek commented above . He tested and found an error.

**Comment by yxun** [comment]
@jewertow I reverted the configmap data entry name change. If you have a running cluster, could you check it now ?
I am waiting on my cluster and then I will verify it again. Thanks.

**Comment by jewertow** [comment]
Ok, let me test it.

**Comment by jewertow** [comment]
@yxun still the same error...

**Comment by jewertow** [comment]
When I installed SMCP v2.2, then DaemonSet `istio-cni-node-v2-3` started to work.

I noticed that `istio-cni-node-v2-3` did not evaluate `__LOG_UDS_ADDRESS__` and `__KUBECONFIG_FILEPATH__` - it logs:
```
2022-10-04T17:34:25.629064Z     info    CNI install configuration:
CNINetDir: /etc/cni/multus/net.d
MountedCNINetDir: /host/etc/cni/multus/net.d
CNIConfName: v2-3-istio-cni.conf
ChainedCNIPlugin: false
CNINetworkConfigFile:
CNINetworkConfig: {
  "cniVersion": "0.3.1",
  "name": "v2-3-istio-cni",
  "type": "v2-3-istio-cni",
  "log_level": "",
  "log_uds_address": "__LOG_UDS_ADDRESS__",
  "kubernetes": {
      "kubeconfig": "__KUBECONFIG_FILEPATH__",
      "cni_bin_dir": "/opt/multus/bin",
      "exclude_namespaces": [ "istio-operator" ]
  }
}
CNIEnableInstall: true
CNIEnableReinstall: true
LogLevel: warn
KubeconfigFilename: v2-3-istio-cni.kubeconfig
KubeconfigMode: 0600
KubeCAFile:
SkipTLSVerify: false
K8sServiceProtocol:
K8sServiceHost: 10.217.4.1
K8sServicePort: 443
K8sNodeName: istio-cni-node-v2-3-rrp4t
UpdateCNIBinaries: true
SkipCNIBinaries: []
MonitoringPort: 15014
LogUDSAddress: /var/run/istio-cni/log.sock
```
I didn't remove failing daemon set and just updated SMCP version to `v2.2`. Then both daemon sets got up and running, but logs contain evaluated paths:
```
2022-10-04T17:58:33.912648Z	info	install	CNI config: {
  "cniVersion": "0.3.1",
  "name": "v2-3-istio-cni",
  "type": "v2-3-istio-cni",
  "log_level": "",
  "log_uds_address": "/var/run/istio-cni/log.sock",
  "kubernetes": {
      "kubeconfig": "/etc/cni/multus/net.d/v2-3-istio-cni.kubeconfig",
      "cni_bin_dir": "/opt/multus/bin",
      "exclude_namespaces": [ "istio-operator" ]
  }
}
```

**Comment by yxun** [comment]
I can reproduce this error. The only difference in the configmap data. In v2.3, we don't mention any iptable executable or script.

**Comment by yxun** [comment]
It looks like a race condition in the initial v2.3 SMCP creation.
When I deleted the previous SMCP, daemonset and configmap, the next run of v2.3 SMCP creation works fine and all v2-3 daemonset pods running fine now.
I will check the NewManifestProcessor code and add a change if needed.

**Comment by luksa** [comment]
> I can reproduce this error. The only difference in the configmap data. In v2.3, we don't mention any iptable executable or script.

In 2.3 the executable is built into the istio-cni plugin, that's why it does need to be mentioned.

**Comment by yxun** [comment]
The correct log from a running cni-v2-3 daemonset pod

```
2022-10-04T21:53:40.995476Z info Start a UDS server for CNI plugin logs
2022-10-04T21:53:40.995536Z info repair Start CNI race condition repair.
2022-10-04T21:53:41.000283Z info ControlZ available at 127.0.0.1:9876
2022-10-04T21:53:41.109696Z info install Copied istio-cni to /host/opt/cni/bin/v2-3-istio-cni.
2022-10-04T21:53:41.109729Z info install Directory /host/secondary-bin-dir is not writable, skipping.
2022-10-04T21:53:41.109922Z info install write kubeconfig file /host/etc/cni/multus/net.d/v2-3-istio-cni.kubeconfig with:
# Kubeconfig file for Istio CNI plugin.
apiVersion: v1
kind: Config
...
```

The error cni pod log from a clean cluster initial run

```
2022-10-04T14:41:56.694363Z	error	Installer exits with open /host/etc/cni/multus/net.d/v2-3-istio-cni.kubeconfig.tmp.3542813523: no such file or directory
2022-10-04T14:41:56.694369Z	info	install	Cleaning up.
2022-10-04T14:41:56.694378Z	info	install	Removing binary: /host/opt/cni/bin/v2-3-istio-cni
Error: open /host/etc/cni/multus/net.d/v2-3-istio-cni.kubeconfig.tmp.3542813523: no such file or directory
```

"repair Start CNI race condition repair."  There is a race condition and it triggers a StartRepair step in :
https://github.com/maistra/istio/blob/maistra-2.3/cni/pkg/repair/repair.go#L226

**Comment by yxun** [comment]
> __KUBECONFIG_FILEPATH__  and __LOG_UDS_ADDRESS__
We didn't set those two in our cni.go renderCNI function. In previous releases, they were the same place holders in the configmap

**Comment by luksa** [comment]
> I noticed that `istio-cni-node-v2-3` did not evaluate `__LOG_UDS_ADDRESS__` and `__KUBECONFIG_FILEPATH__` - it logs:

@jewertow Are you saying that the actual config file contained these two placeholders or that you saw them only in the log? It's normal for this to be in the log statement that reads `CNINetworkConfig: {`. The CNI installer running in the DaemonSet is the one that replaces the placeholders with actual values when it writes the file (the file contents are printed in the log statement that reads `install	CNI config: {`. 

So, there's no issue here.

**Comment by luksa** [comment]
@yxun, @jewertow I can also reproduce this. I know what the problem is and I'll have a fix ready in a few minutes.

**Comment by luksa** [comment]
The fix for the issue is in https://github.com/maistra/istio/pull/638

**Comment by luksa** [comment]
@yxun You can change the config map entry key back to `cni_network_config`. And while you're at it, I suggest also changing the container name in the 2.3 DaemonSet from `install-cni-v2-3` to just `install-cni`.

**Comment by luksa** [comment]
@yxun okay to merge?

**Comment by yxun** [comment]
> @yxun okay to merge?

ok

**Pull request merged**
